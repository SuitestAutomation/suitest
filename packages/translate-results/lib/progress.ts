import t from './texts';
import {Translation} from './types';
import {StartupError, translateStartupError} from './startup';

/**
 * @description Get humanized progress status explanation result
 */
export function translateProgress(message: ProgressMessage): Translation {
	if ('code' in message) {
		return translateStartupError(message.code);
	}

	switch (message.status) {
		case 'openingApp': return {title: t['progress.status.openingApp']()};
		case 'closingApp': return {title: t['progress.status.closingApp']()};
		case 'bootingDevice': return {title: t['progress.status.bootingDevice']()};
		case 'needManual': return {title: t['progress.status.needManual']()};
		case 'recoveringID': return {title: t['progress.status.recoveringID']()};
		case 'waitingForConnectionFromBootstrap': return {title: t['progress.status.waitingForConnectionFromBootstrap']()};
		case 'waitingForConnectionFromIL': return {title: t['progress.status.waitingForConnectionFromIL']()};
		case 'unistallingApp': return {title: t['progress.status.unistallingApp']()};
		case 'uploadingAndInstallingApp': return {title: t['progress.status.uploadingAndInstallingApp']()};
		case 'nothing': break;
		// actionFailed should comes with code, so it will proceeded by translateStartupError above
		case 'actionFailed': break;
		default:
			const _status: never = message.status;
			console.warn(_status, 'progression status not handled');
	}

	return {title: ''};
}

export const PROGRESS_STATUS = Object.freeze({
	actionFailed: 'actionFailed',
	openingApp: 'openingApp',
	closingApp: 'closingApp',
	bootingDevice: 'bootingDevice',
	needManual: 'needManual',
	nothing: 'nothing',
	recoveringID: 'recoveringID',
	waitingForConnectionFromBootstrap: 'waitingForConnectionFromBootstrap',
	waitingForConnectionFromIL: 'waitingForConnectionFromIL',
	unistallingApp: 'unistallingApp',
	uploadingAndInstallingApp: 'uploadingAndInstallingApp',
} as const);

export type ProgressStatus = keyof typeof PROGRESS_STATUS;

type ProgressMessage = {
	status?: typeof PROGRESS_STATUS.actionFailed
	code: StartupError,
} | {
	status: ProgressStatus,
};
