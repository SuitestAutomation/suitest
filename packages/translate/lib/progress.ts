import t from './texts';
import {Translation} from './types';
import {PROGRESS_STATUS, ProgressStatus, NotStartedReason} from './constants';
import {translateNotStartedReason} from './notStartedReason';

/**
 * @description Get humanized progress status explanation result
 */
export function translateProgress(message: ProgressMessage): Translation | undefined {
	if ('code' in message && message.code !== undefined) {
		return translateNotStartedReason(message.code);
	}

	switch (message.status) {
		case PROGRESS_STATUS.OPENING_APP: return {title: t['progress.status.openingApp']()};
		case PROGRESS_STATUS.CLOSING_APP: return {title: t['progress.status.closingApp']()};
		case PROGRESS_STATUS.BOOTING_DEVICE: return {title: t['progress.status.bootingDevice']()};
		case PROGRESS_STATUS.WAIT_FOR_MANUAL_ACTION: return {title: t['progress.status.needManual']()};
		case PROGRESS_STATUS.DEVICE_IDENTIFICATION: return {title: t['progress.status.recoveringID']()};
		case PROGRESS_STATUS.WAITING_FOR_BOOTSTRAP: return {title: t['progress.status.waitingForConnectionFromBootstrap']()};
		case PROGRESS_STATUS.WAITING_FOR_IL: return {title: t['progress.status.waitingForConnectionFromIL']()};
		case PROGRESS_STATUS.APP_UNINSTALL: return {title: t['progress.status.unistallingApp']()};
		case PROGRESS_STATUS.APP_UPLOAD_INSTALL: return {title: t['progress.status.uploadingAndInstallingApp']()};
		case PROGRESS_STATUS.NOTHING: return {title: ''};
		// actionFailed should come with code, so it will be proceeded by translateStartupError above
		case 'actionFailed': break;
		default:
			const _status: never = message.status;
			console.warn(_status, 'progress status not handled');
	}

	return undefined;
}

type ProgressMessage = {
	status?: typeof PROGRESS_STATUS.ACTION_FAILED,
	code: NotStartedReason,
} | {
	status: ProgressStatus,
	code?: undefined,
};
