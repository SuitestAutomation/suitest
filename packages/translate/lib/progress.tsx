import {Translation} from './types';
import {PROGRESS_STATUS, ProgressStatus, NotStartedReason} from './constants';
import {translateNotStartedReason} from './notStartedReason';
import {jsx} from '@suitest/smst';

/**
 * @description Get humanized progress status explanation result
 * @throws {Error} Throws an Error if unknown status is provided or code is required and missing
 */
export function translateProgress(message: ProgressMessage): Translation {
	if ('code' in message && message.code !== undefined) {
		return translateNotStartedReason(message.code);
	}

	if (message.status === undefined) {
		throw new Error('A status is expected.');
	}

	switch (message.status) {
		case PROGRESS_STATUS.OPENING_APP: return {
			title: <text>Trying to open app…</text>,
		};
		case PROGRESS_STATUS.CLOSING_APP: return {
			title: <text>Trying to close app…</text>,
		};
		case PROGRESS_STATUS.BOOTING_DEVICE: return {
			title: <text>Running the boot sequence defined for the device…</text>,
		};
		case PROGRESS_STATUS.WAIT_FOR_MANUAL_ACTION: return {
			title: <text>Paused. For this platform, install and open the application manually.</text>,
		};
		case PROGRESS_STATUS.DEVICE_IDENTIFICATION: return {
			title: <text>Trying to recover Suitest device ID…</text>,
		};
		case PROGRESS_STATUS.SELECTING_SUITESTIFY_INSTANCE: return {
			title: <text>Selecting nearest suitestify instance…</text>,
		};
		case PROGRESS_STATUS.WAITING_FOR_BOOTSTRAP: return {
			title: <text>Waiting for connection from the Suitest app on device…</text>,
		};
		case PROGRESS_STATUS.WAITING_FOR_IL: return {
			title: <text>Waiting for connection from the instrumentation library…</text>,
		};
		case PROGRESS_STATUS.APP_UNINSTALL: return {
			title: <text>Uninstalling app…</text>,
		};
		case PROGRESS_STATUS.APP_UPLOAD_INSTALL: return {
			title: <text>Uploading and installing app…</text>,
		};
		case PROGRESS_STATUS.NOTHING:
		case PROGRESS_STATUS.ACTION_FAILED: return {
			title: <text>{''}</text>,
		};
		default:
			const _status: never = message.status;
			throw new Error(`Unknown status code received: ${_status}`);
	}
}

type ProgressMessage = {
	code?: NotStartedReason,
	status?: ProgressStatus,
};
