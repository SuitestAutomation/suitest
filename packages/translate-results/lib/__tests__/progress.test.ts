import {translateProgress} from '../progress';

describe('Interactive progress explanation.', () => {
	describe('Test actionFailed status with codes.', () => {
		it('Should return empty result if code not provided', () => {
			expect(translateProgress({status: 'actionFailed'}))
				.toStrictEqual({title: ''});
		});

		it('Should return empty result if code is unknown', () => {
			expect(translateProgress({status: 'actionFailed', code: 'unknownCode' as any}))
				.toStrictEqual({title: ''});
		});

		it('Should return translation if code is valid', () => {
			expect(translateProgress({code: 'noActivePlan'}))
				.toStrictEqual({
					title: 'Cannot continue: Your subscription has expired',
					description: 'Your subscription has expired, to continue using Suitest please [renew your subscription](https://the.suite.st/preferences/billing)',
				});
		});
	});

	describe('Test results with', () => {
		it('status "nothing" should be empty string.', () => {
			expect(translateProgress({status: 'nothing'}))
				.toStrictEqual({title: ''});
		});

		it('unknown status should be empty string.', () => {
			expect(translateProgress({status: 'unknownStatus' as any}))
				.toStrictEqual({title: ''});
		});

		it('valid status should return translation.', () => {
			expect(translateProgress({status: 'bootingDevice'}))
				.toStrictEqual({title: 'Running the boot sequence defined for the device...'});
		});
	});
});
