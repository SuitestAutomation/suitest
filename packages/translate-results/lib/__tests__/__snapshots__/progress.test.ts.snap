// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interactive progress explanation. Translation result for "actionFailed" status 1`] = `
Object {
  "title": "",
}
`;

exports[`Interactive progress explanation. Translation result for "bootingDevice" status 1`] = `
Object {
  "title": "Running the boot sequence defined for the device...",
}
`;

exports[`Interactive progress explanation. Translation result for "closingApp" status 1`] = `
Object {
  "title": "Trying to close app...",
}
`;

exports[`Interactive progress explanation. Translation result for "needManual" status 1`] = `
Object {
  "title": "Paused. For this platform, install and open the application manually.",
}
`;

exports[`Interactive progress explanation. Translation result for "nothing" status 1`] = `
Object {
  "title": "",
}
`;

exports[`Interactive progress explanation. Translation result for "openingApp" status 1`] = `
Object {
  "title": "Trying to open app...",
}
`;

exports[`Interactive progress explanation. Translation result for "recoveringID" status 1`] = `
Object {
  "title": "Trying to recover Suitest device ID...",
}
`;

exports[`Interactive progress explanation. Translation result for "unistallingApp" status 1`] = `
Object {
  "title": "Uninstalling app...",
}
`;

exports[`Interactive progress explanation. Translation result for "uploadingAndInstallingApp" status 1`] = `
Object {
  "title": "Uploading and installing app...",
}
`;

exports[`Interactive progress explanation. Translation result for "waitingForConnectionFromBootstrap" status 1`] = `
Object {
  "title": "Waiting for connection from the Suitest app on device...",
}
`;

exports[`Interactive progress explanation. Translation result for "waitingForConnectionFromIL" status 1`] = `
Object {
  "title": "Waiting for connection from the instrumentation library...",
}
`;
